ui/raidboss/data/06-ew/raid/p12s.ts

```diff
@@ -604,30 +638,179 @@ const triggerSet: TriggerSet<Data> = {
+          } else if (x < 100 && y < 80) { // x = 97 && y = 75
+            data.engravement1BeamsPosMap.set('SW', color);
+          } else if (x > 100 && y < 80) { // x = 103 && y = 75
+            data.engravement1BeamsPosMap.set('SE', color);
+          } else if (x > 120 && y < 100) { // x = 125 && y = 97
+            data.engravement1BeamsPosMap.set('NW', color);
+          } else if (x > 120 && y > 100) { // x = 125 && y = 103
+            data.engravement1BeamsPosMap.set('SW', color);
+          } else if (x > 100 && y > 120) { // x = 103 && y = 125
+            data.engravement1BeamsPosMap.set('NE', color);
+          } else if (x < 100 && y > 120) { // x = 97 && y = 125
+            data.engravement1BeamsPosMap.set('NW', color);
+          } else if (x < 80 && y > 100) { // x = 75 && y = 103
+            data.engravement1BeamsPosMap.set('SE', color);
+          }
+        } else if (data.triggerSetConfig.engravement1DropTower === 'tilt') {
+          if (x < 80 && y < 100) { // x = 75 && y = 97
+            data.engravement1BeamsPosMap.set('SE', color);
+          } else if (x < 100 && y < 80) { // x = 97 && y = 75
+            data.engravement1BeamsPosMap.set('SE', color);
+          } else if (x > 100 && y < 80) { // x = 103 && y = 75
+            data.engravement1BeamsPosMap.set('SW', color);
+          } else if (x > 120 && y < 100) { // x = 125 && y = 97
+            data.engravement1BeamsPosMap.set('SW', color);
+          } else if (x > 120 && y > 100) { // x = 125 && y = 103
+            data.engravement1BeamsPosMap.set('NW', color);
+          } else if (x > 100 && y > 120) { // x = 103 && y = 125
+            data.engravement1BeamsPosMap.set('NW', color);
+          } else if (x < 100 && y > 120) { // x = 97 && y = 125
+            data.engravement1BeamsPosMap.set('NE', color);
+          } else if (x < 80 && y > 100) { // x = 75 && y = 103
+            data.engravement1BeamsPosMap.set('NE', color);
+          }
+        }
+      },
+    },
+    {
+      id: 'P12S Engravement 1 Beam',
+      type: 'StartsUsing',
+      netRegex: { id: ['82F[12]'], source: 'Anthropos' },
+      condition: (data) => data.engravementCounter === 1,
+      alertText: (data, matches, output) => {
+        if (data.me === matches.target) {
+          if (matches.id === '82F1')
+            return output.lightBeam!();
+          return output.darkBeam!();
+        }
+      },
+      outputStrings: {
+        lightBeam: {
+          en: 'light beam',
+          cn: '引导光激光',
+        },
+        darkBeam: {
+          en: 'dark beam',
+          cn: '引导暗激光',
+        },
+      },
+    },
     {
       id: 'P12S Engravement 1 Tower Drop',
       type: 'GainsEffect',
       netRegex: { effectId: engravementTowerIds },
       condition: (data) => data.engravementCounter === 1,
-      durationSeconds: (_data, matches) => parseFloat(matches.duration),
+      delaySeconds: 0.1, // Make sure Drop Tower Pos Tracker is done
```

Comment:
If I look at the log times, I see this:
```
// Anthropos move into position
[19:12:18.911] 261 105:Change:4001EA0D:Heading:0.0000:PosX:97.0000:PosY:75.0000:PosZ:0.0000
// tethers appear
[19:12:22.471] Tether 23:4001EA0D:Anthropos:etc
// P12S Engravement 1 Tower Drop + 0.1s delay
[19:12:22.471] StatusAdd 1A:DFC:etc
// P12S Engravement 1 Drop Tower Pos Tracker + async getCombatants
[19:12:22.516] StartsCasting 14:4001EA0D:Anthropos:82F1:etc
```

Sorry, it's the delay here that makes me a little nervous.  It's possible that there might be some local (or network, if this is a remote overlay) delay in getting combatant data, and so you're hoping that 0.1s is longer than 0.04s + the time it takes to do 4x calls to `getCombatants`.  This seems like risky timing to me.

My original suggestion was this:
* add a new Anthropos collector trigger that records all the ids of any Anthropos tethers
* assume that 1A gains effect lines come after 23 tether lines always (this seems consistent to me)
* P12S Engravement 1 Tower Drop trigger then waits for the GainsEffect on the player, and does a getCombatants with the Anthropos ids, and does the work of "P12S Engravement 1 Drop Tower Pos Tracker" and "P12S Engravement 1 Tower Drop" in a single trigger

...that way there's no need for a heuristic delay (that might be wrong) and the call comes out as soon as the information about the combatants is ready.

Reply:
This looks more reasonable, let me see how to implement it.

Reply:
Change P12S Engravement 1 Drop Tower Pos Tracker to the `[19:12:22.471] Tether 23:4001EA0D:Anthropos:etc`, Because you need to capture the locations of the two Tether, I just handle them separately.

https://github.com/quisquous/cactbot/pull/5597/commits/095f49e847fa5c465dfd2f4cd63a209223827cee

Reply:
I'm sorry, but I still think this is a race.  This is now:
```

Reply:
I'm sorry but I still think this is a race.  If I am reading this correctly, this is now:

```
// Anthropos move into position
[19:12:18.911] 261 105:Change:4001EA0D:Heading:0.0000:PosX:97.0000:PosY:75.0000:PosZ:0.0000
// P12S Engravement 1 Drop Tower Pos Tracker + async getCombatants
[19:12:22.471] Tether 23:4001EA0D:Anthropos:etc
// P12S Engravement 1 Tower Drop (hope that the async calls finish?)
[19:12:22.471] StatusAdd 1A:DFC:etc
[19:12:22.516] StartsCasting 14:4001EA0D:Anthropos:82F1:etc
```

There are 4 async calls there that need to be done instantaneously for the StatusAdd line to have the information there?

I'm trying to propose this structure:

`P12S Engravement 1 Tether Tracker`
* a new trigger, on the tether
* records all of the tether `matches.sourceId` into some new `data.engravement1TetherIds`

`P12S Engravement 1 Tower Drop` (combine logic from `Tower Drop` and `Tower Pos Tracker`)
* condition: only run if the status effect is on you
* promise: call get combatants with all of the `data.engravement1TetherIds`
* alertText:
  * for each combatant, do the logic from `P12S Engravement 1 Drop Tower Pos Tracker` to set the `data.engravement1BeamsPosMap`
  * then do the rest of the normal logic for Tower Drop to use that map

The timing in the PR right now is:
* tethers, call 4 async functions
* immediately after, status effects, hope async functions are done??

This timing is:
* tethers collect ids synchronously
* immediately after, status effects call 1 async function, alert text will run naturally when async function is done (no matter when it completes)

Does that make sense?

Reply:
This is a good method, I did not know that getCombatants can query multiple data, so I wrote it separately. but for now this is great way to deal with it.

https://github.com/quisquous/cactbot/pull/5597/commits/b60f413ff26c4c5ee41972a46e2f801fb43d77f3

Reply:
This is a good method, I didn't know that getCombatants can take multiple objects at once, so writing them separately caused the problem, Sorry for that.

https://github.com/quisquous/cactbot/pull/5597/commits/b60f413ff26c4c5ee41972a46e2f801fb43d77f3

Reply:
Thank you, this is perfect.

For what it's worth, I think the "multiple objects" part is minor, but it's the asynchronous delay that is the major issue.  If the asynchronous call took (for example) an entire second to complete for some reason (weird network delays, very slow computer, bugs) then the previous triggers would not work because the `Tower Drop` trigger would run before `Tower Pos Tracker` had set any values in the map (because the asynchronous function hadn't completed yet).  With this latest PR change, no matter how long the asynchronous function takes, the trigger logic will still work.